R/teal_lockfile.R

```diff
@@ -0,0 +1,94 @@
+#' Create application `.lockfile`
+#'
+#' This function is used during `teal::init` and enables to create `renv` `.lockfile` to be shared in the `teal` app.
+#'
+#' The process uses [renv::snapshot()], which allows multiple ways of `.lockfile` creation.
+#'
+#' - **Pre-computed `.lockfile`**: User is able to provide his own pre-computed `.lockfile` by setting the path to the
+#' `.lockfile` through `options(teal.renv.lockfile = "")`. Then this function is not used.
+#' - **Working directory `.lockfile`**: If `options(teal.renv.lockfile)` is empty, by default `teal` will
+#' create an `implicit` type of the `.lockfile`, that uses `renv::dependencies()` to detect all R packages in the
+#' current project working directory.
+#' - **`DESCRIPTION` based `.lockfile`**: You can always include a `DESCRIPTION` file in your working directory and
+#' enable `.lockfile` creation based on this file. To do this, run `renv::settings$snapshot.type("explicit")`.
+#' Naming of `type` is the same as in `renv::snapshot()`. For the `"explicit"` type refer to
+#' `renv::settings$package.dependency.fields()`
+#' to see what `DESCRIPTION` fields are included in the `.lockfile`.
+#' - **Custom files based `.lockfile`**: If you want to specify custom files as a base for the `.lockfile`, then run
+#' `renv::settings$snapshot.type("custom")` and set `renv.snapshot.filter` option.
+#'
+#' @note
+#' This function computes the `.lockfile` as a `promises::future_promise` promise, while
+#' running the evaluation of the process on a separate worker. If `future::plan()` was set to something different than
+#' `future::sequential`, it reuses the parallel backed. If not, it sets `future::multisession` as a parallel plan with
+#' 2 workers. Then, `shiny::ExtendedTask()` is used to run asynchronous computations. If the `teal` app started with
+#' `future::sequential` plan, is it set back once the task is finished.
+#'
+#' @section `.lockfile` usage:
+#' Once you have a `.lockfile` file, you can restore R project with `ren::init();renv::restore()`.
+#'
+#' @seealso [renv::snapshot()], [renv::restore()].
+#'
+#' @return Nothing is returned. The side effect of this function is the creation of
+#' `getOption("teal.internal.renv.lockfile")` file, which will be shared through the `teal` app.
+#'
+#' @keywords internal
+teal_lockfile <- function() {
+  # If user has setup the file, there is no need to compute a new one.
+  user_lockfile <- getOption("teal.renv.lockfile", "")
+  if (!identical(user_lockfile, "") && !file.exists(user_lockfile)) {
+    stop("lockfile provided through options('teal.renv.lockfile') does not exist.")
+  }
```

Comment:
One more thing bothers me. Right now we have two separate locations to check - the one with user provided file and the second one created internally (right now stored in option, we can hardcode - it's irrelevant here). Assuming you handle edge cases here - can we exit the function with just one path to check? Then two (or multiple paths) are handled in _create_ and _read_ will be having just one - so essentially two (or multiple paths) will be implemented in one place only. That would be simpler I believe.

Reply:
So there are two paths because
- user can provide his/her pre-computed lockfile and then we dont start the parallel background

We use two paths, because we do not return (anymore) anything through `teal_lockfile` function, so that we do not pass anything from `teal::init` through `teal::srv_teal_with_splash` through `teal::srv_teal`. If we allow the extension of parameters through `teal::init` -> `teal::srv_teal_with_splash` -> `teal::srv_teal` so that a path is provided, then we can handle one path in `teal_lockfile` and no need for further checks. But we didn't want to introduce any parameters in `teal::srv_teal`, hence there is no way what happened in `teal_lockfile`, so we do check both paths in `downloadHandler`

Reply:
how about this:
```
return_path <- "some/hard/coded/path.lock
if (user_provided()) {
  file.move(user_lockfile, return_path)
} else {
  file_create(return_path)
}
return(invisible(NULL))
```

Reply:
👍

Reply:
Hey, that was a great simplification a8d656b6ef79c7a60824d709a658d1b0fb8b4056